_posts/2024-04-12-learn-from-the-xz-backdoor.md

```diff
@@ -0,0 +1,111 @@
+
+
+## Analysis of the backdoor
+
+Analysis of the object code turned out to be time intensive. While the 2nd stage that checked for the right build conditions (is it a distribution build, does it have the expected compiler environment etc) was easy to decode and helped us understand the potential impact, initially it wasn't really clear to us what the obfuscated object code that was injected during the build was doing.
+
+By using Ghidra, we were able to get somewhat readable C code back from the injected machine code, so we started trying to decipher the puzzle of code. Spotting the entry point in the `_get_cpuid` function that was part of the IFUNC handling was one of the first findings. Just googling these combination of words led to an upstream discussion, to the disablement of ifunc in the oss fuzz project and an interesting bugreport in the Fedora community where valgrind issues were reported with XZ 5.6.0 and apparently the upstream was fixing those by updating unrelated things including "the test files" in the repository. So not only commits in the repository but also misleading communication around issue directly related to those commits made it obvious that we were not finding an unfortunate accident by an innocent maintainer who might have been hacked, but a planned action by the current upstream maintainer. Just in case the alarm bells weren't loud enough already, this doubled their noise level.
+
+
+## Preparing for the Public Disclosure
+
+Combining all of what we learned so far, the picture became clearer. Somebody has spent years of preparation to lay down the ground work, build up a good maintainer reputation, take over the project and then chose a point in time that was a critical time window for several distribution projects and in the middle of Lunar New Year as well as other holidays to release a new release with new features and an obfuscated backdoor that was well crafted to target only specific distributions, namely those using GCC, Binutils, Glibc with RPM or Debian build processes on x86_64.
+
+With all of that in mind, we realized that there is going to be a lot of public coverage on this. It will be in the news for days to weeks. So we started a new workstream to prepare for that with the comms teams.
+
+
+## Public Disclosure
+
+By the time of the public disclosure, all workstreams had already completed. We identifed the list of affected products, and had already released all updates for all affected ones. Communication was ready to be put online and sent out to the relevant parties. All of that was possible because many people went above and beyond, put everything else aside to react timely and with a lot of engagement to ensure we haven't missed anything or overlooked anything, all while a long public holidays weekend had already started. Kudos to everyone who was working around the clock on preparing for this.
```

Comment:
```suggestion
By the time of the public disclosure, all workstreams had already completed. We identifed the list of affected products, and had already released all updates for all affected ones. Communication was ready to be put online and sent out to the relevant parties. All of that was possible because many people went above and beyond, put everything else aside to react timely and with a lot of engagement to ensure we hadn't missed or overlooked anything, all while a long public holiday weekend had already started. Kudos to everyone who was working around the clock on preparing for this.
```